runtime/signals.c

```diff
@@ -29,101 +28,144 @@
+    if (atomic_load_explicit(&caml_pending_signals[i], memory_order_acquire)) return 1;
+  }
+  return 0;
+}
 
 /* Execute all pending signals */
 
-static void caml_execute_signal(int signal_number);
-void caml_process_pending_signals(void)
+CAMLexport value caml_process_pending_signals_exn(void)
 {
   int i;
-  /* this function preserves errno (PR#5982) */
-  int saved_errno = errno;
-
-  if (caml_signals_are_pending) {
-    caml_signals_are_pending = 0;
-    for (i = 0; i < NSIG; i++) {
-      if (caml_pending_signals[i]) {
-        caml_pending_signals[i] = 0;
-        caml_execute_signal(i);
+  intnat specific_signal_pending;
+  value exn;
+#ifdef POSIX_SIGNALS
+  sigset_t set;
+#endif
+
+if( atomic_load_explicit(&total_signals_pending, memory_order_acquire) == 0 )
+  return Val_unit;
+
+/* Check that there is indeed a pending signal before issuing the
+    syscall in [pthread_sigmask]. It is possible for there to be a
+    race between checking [total_signals_pending] and checking the 
+    actual pending signals so there may not be a signal to handle. */
+if (!check_for_pending_signals())
+  return Val_unit;
+
+#ifdef POSIX_SIGNALS
+  pthread_sigmask(/* dummy */ SIG_BLOCK, NULL, &set);
+#endif
+  for (i = 0; i < NSIG; i++) {
+    if ( atomic_load_explicit(&caml_pending_signals[i], memory_order_acquire) == 0 )
+      continue;
+#ifdef POSIX_SIGNALS
+    if(sigismember(&set, i))
+      continue;
+#endif
+  again:
+    specific_signal_pending = atomic_load_explicit(&caml_pending_signals[i], memory_order_acquire);
+    if( specific_signal_pending > 0 ) {
+      if( !atomic_compare_exchange_strong(
+            &caml_pending_signals[i], 
+             &specific_signal_pending, specific_signal_pending - 1) ) {
+        /* We failed our CAS because another thread beat us to processing
+           this signal. Try again to see if there are more of this signal
+           to process. */
+        goto again;
       }
+
+      atomic_fetch_sub_explicit(&total_signals_pending, 1, memory_order_acq_rel);
+
+      exn = caml_execute_signal_exn(i, 0);
+      if (Is_exception_result(exn)) return exn;
     }
   }
-  errno = saved_errno;
+  return Val_unit;
 }
 
 /* Record the delivery of a signal, and arrange for it to be processed
    as soon as possible:
-   - in bytecode: via caml_something_to_do, processed in caml_process_event
-   - in native-code: by playing with the allocation limit, processed
-       in caml_garbage_collection
+   - via total_signals_are_pending, processed in
+     caml_process_pending_signals_exn.
+   - by playing with the allocation limit, processed in
+     caml_garbage_collection
 */
 
+CAMLno_tsan
 CAMLexport void caml_record_signal(int signal_number)
 {
-  caml_pending_signals[signal_number] = 1;
-  caml_signals_are_pending = 1;
+  atomic_fetch_add_explicit(&caml_pending_signals[signal_number], 1, memory_order_seq_cst);
+  atomic_fetch_add_explicit(&total_signals_pending, 1, memory_order_seq_cst);
```

Comment:
I am not an expert on memory ordering. Can you explain why we do `memory_order_acq_rel` at line 98 [1] but `memory_order_seq_cst` here? I wonder whether adding comments on the choice of memory ordering should be attached to the declaration of atomic variables, if they use interesting memory ordering. For `total_signals_pending`, we may mention that the we use acquire for loads and sequential consistency for CASes?

[1] https://github.com/ocaml-multicore/ocaml-multicore/pull/630/files#diff-93ee9a0b40685be35da1909f57cba82cfe51d5a8de495947e5068a8b8d8172cbR98

Reply:
That the other ordering is acquire release is an oversight. I went with Tom's suggestion of going for the strongest possible ordering and we can look at loosening them up if they proved to be a problem.

I've moved the other `signals.c` operations to `seq_cst` now.